File: src/sc_neurocore/neurons/models/adaptive_threshold_moe.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Adaptive Threshold MoE Neuron (SpikingBrain)

"""Adaptive threshold neuron with integer spike counts for Mixture-of-Experts.

Implements the SpikingBrain activation function (arXiv:2509.05276v2, Sep 2025)
which produces *integer* spike counts rather than binary spikes, enabling
direct use in transformer MoE layers with minimal information loss.

Equations (SpikingBrain-1.0, Section 3.2):

    V_th = (1/k) · mean(|x|)          — adaptive threshold from input statistics
    v[t+1] = v[t] + x[t+1]            — membrane integration
    s_INT = round(v / V_th)            — integer spike count
    v <- v - V_th · s                  — soft reset (residual preserved)

Time-collapsed mode (single-step inference):

    s_INT = round(x / V_th)

Parameter *k* controls the firing rate / sparsity trade-off:
higher k -> lower threshold -> more spikes -> less sparsity.

Reference: SpikingBrain-1.0, arXiv:2509.05276v2, September 2025.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field


@dataclass
class AdaptiveThresholdMoENeuron:
    """Adaptive threshold MoE spiking neuron (SpikingBrain).

    Parameters
    ----------
    k : float
        Firing rate control (higher k -> lower threshold -> more spikes).
        Default: 4.0 (SpikingBrain recommended).
    ema_alpha : float
        EMA decay for running mean of |input|. Default: 0.1.
    """

    k: float = 4.0
    ema_alpha: float = 0.1

    v: float = field(default=0.0, repr=False)
    v_th: float = field(default=1.0, repr=False)
    _mean_abs_x: float = field(default=0.0, repr=False)

    def __post_init__(self) -> None:
        if not math.isfinite(self.k) or self.k <= 0.0:
            raise ValueError("k must be finite and positive")
        if not math.isfinite(self.ema_alpha) or not (0.0 < self.ema_alpha <= 1.0):
            raise ValueError("ema_alpha must be finite and in (0, 1]")
        if not math.isfinite(self.v):
            raise ValueError("v must be finite")
        if not math.isfinite(self.v_th) or self.v_th <= 0.0:
            raise ValueError("v_th must be finite and positive")
        if not math.isfinite(self._mean_abs_x) or self._mean_abs_x < 0.0:
            raise ValueError("_mean_abs_x must be finite and non-negative")

    def step(self, current: float) -> int:
        """Advance one timestep. Returns integer spike count (>= 0).

        Implements: V_th = mean(|x|)/k, v += x, s = round(v/V_th), v -= V_th*s.
        """
        if not math.isfinite(current):
            raise ValueError("current must be finite")
        self._validate_runtime_state()

        next_mean_abs_x = (1.0 - self.ema_alpha) * self._mean_abs_x + self.ema_alpha * abs(current)
        next_v_th = self._threshold_from_mean(next_mean_abs_x)
        next_v = self.v + current
        if not math.isfinite(next_v):
            raise ValueError("soft reset residual must remain finite")
        ratio = next_v / next_v_th if next_v_th > 1e-12 else 0.0
        if not math.isfinite(ratio):
            raise ValueError("adaptive threshold ratio must remain finite")
        s_int = max(round(ratio), 0)
        residual = next_v - next_v_th * s_int if s_int != 0 else next_v
        if not math.isfinite(residual):
            raise ValueError("soft reset residual must remain finite")

        self._mean_abs_x = next_mean_abs_x
        self.v_th = next_v_th
        self.v = residual
        return s_int

    def step_collapsed(self, activation: float) -> int:
        """Time-collapsed single-step: s_INT = round(x / V_th)."""
        if not math.isfinite(activation):
            raise ValueError("activation must be finite")
        self._validate_runtime_state()

        next_mean_abs_x = (1.0 - self.ema_alpha) * self._mean_abs_x + self.ema_alpha * abs(
            activation
        )
        next_v_th = self._threshold_from_mean(next_mean_abs_x)
        ratio = activation / next_v_th
        if not math.isfinite(ratio):
            raise ValueError("adaptive threshold ratio must remain finite")
        spike_count = max(round(ratio), 0)
        self._mean_abs_x = next_mean_abs_x
        self.v_th = next_v_th
        return spike_count

    def sparsity(self) -> float:
        """Current activation sparsity (1.0 if below threshold, 0.0 if firing)."""
        self._validate_runtime_state()
        return 1.0 if abs(self.v) < self.v_th else 0.0

    def reset(self) -> None:
        """Reset state to initial conditions."""
        self.v = 0.0
        self._mean_abs_x = 0.0
        self.v_th = 1.0

    def _validate_runtime_state(self) -> None:
        if not math.isfinite(self.v):
            raise ValueError("runtime membrane state must be finite")
        if not math.isfinite(self.v_th) or self.v_th <= 0.0:
            raise ValueError("runtime threshold state must be finite and positive")
        if not math.isfinite(self._mean_abs_x) or self._mean_abs_x < 0.0:
            raise ValueError("runtime mean absolute input state must be finite and non-negative")

    def _threshold_from_mean(self, mean_abs_x: float) -> float:
        if not math.isfinite(mean_abs_x) or mean_abs_x < 0.0:
            raise ValueError("adaptive threshold mean must remain finite and non-negative")
        threshold = mean_abs_x / self.k if mean_abs_x > 1e-12 else 1.0
        if not math.isfinite(threshold) or threshold <= 0.0:
            raise ValueError("adaptive threshold must remain finite and positive")
        return threshold

File: src/sc_neurocore/accel/julia/neurons/adaptive_threshold_moe.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia for adaptive_threshold_moe

module AdaptiveThresholdMoeAccel

export step!, step_collapsed, simulate, sparsity, valid, AdaptiveThresholdMoENeuronState

mutable struct AdaptiveThresholdMoENeuronState
    k::Float64
    ema_alpha::Float64
    v::Float64
    v_th::Float64
    _mean_abs_x::Float64
end

function AdaptiveThresholdMoENeuronState()
    AdaptiveThresholdMoENeuronState(4.0, 0.1, 0.0, 1.0, 0.0)
end

function valid(s::AdaptiveThresholdMoENeuronState)
    return all(isfinite, (s.k, s.ema_alpha, s.v, s.v_th, s._mean_abs_x)) &&
        s.k > 0.0 &&
        s.ema_alpha > 0.0 &&
        s.ema_alpha <= 1.0 &&
        s.v_th > 0.0 &&
        s._mean_abs_x >= 0.0
end

function threshold_from_mean(mean_abs_x::Float64, k::Float64)
    if !isfinite(mean_abs_x) || mean_abs_x < 0.0 || !isfinite(k) || k <= 0.0
        throw(DomainError(mean_abs_x, "AdaptiveThresholdMoE adaptive threshold mean must remain finite and non-negative"))
    end
    v_th = (mean_abs_x > 1e-12) ? mean_abs_x / k : 1.0
    if !isfinite(v_th) || v_th <= 0.0
        throw(DomainError(v_th, "AdaptiveThresholdMoE adaptive threshold must remain finite and positive"))
    end
    return v_th
end

function step_collapsed(s::AdaptiveThresholdMoENeuronState, activation)
    if !isfinite(activation)
        throw(DomainError(activation, "AdaptiveThresholdMoE activation must be finite"))
    end
    if !valid(s)
        throw(DomainError(s.v, "AdaptiveThresholdMoE runtime state must be finite and physically valid"))
    end
    next_mean_abs_x = (1.0 - s.ema_alpha) * s._mean_abs_x + s.ema_alpha * abs(activation)
    next_v_th = threshold_from_mean(next_mean_abs_x, s.k)
    ratio = activation / next_v_th
    if !isfinite(ratio)
        throw(DomainError(ratio, "AdaptiveThresholdMoE threshold ratio must remain finite"))
    end
    s_int = max(round(ratio), 0)
    s._mean_abs_x = next_mean_abs_x
    s.v_th = next_v_th
    return s_int
end

function sparsity(s::AdaptiveThresholdMoENeuronState)
    if !valid(s)
        throw(DomainError(s.v, "AdaptiveThresholdMoE runtime state must be finite and physically valid"))
    end
    return (abs(s.v) < s.v_th) ? 1.0 : 0.0
end

function step!(s::AdaptiveThresholdMoENeuronState, I_ext::Float64=0.0; dt::Float64=0.1)
    if !isfinite(I_ext)
        throw(DomainError(I_ext, "AdaptiveThresholdMoE input current must be finite"))
    end
    if !valid(s)
        throw(DomainError(s.v, "AdaptiveThresholdMoE runtime state must be finite and physically valid"))
    end
    next_mean_abs_x = (1.0 - s.ema_alpha) * s._mean_abs_x + s.ema_alpha * abs(I_ext)
    next_v_th = threshold_from_mean(next_mean_abs_x, s.k)
    next_v = s.v + I_ext
    if !isfinite(next_v)
        throw(DomainError(next_v, "AdaptiveThresholdMoE soft reset residual must remain finite"))
    end
    ratio = next_v / next_v_th
    if !isfinite(ratio)
        throw(DomainError(ratio, "AdaptiveThresholdMoE threshold ratio must remain finite"))
    end
    s_int = max(round(ratio), 0)
    residual = (s_int != 0) ? next_v - next_v_th * s_int : next_v
    if !isfinite(residual)
        throw(DomainError(residual, "AdaptiveThresholdMoE soft reset residual must remain finite"))
    end
    s._mean_abs_x = next_mean_abs_x
    s.v_th = next_v_th
    s.v = residual
    return s_int
end

function simulate(n_steps::Int=1000; I_ext::Float64=10.0, dt::Float64=0.1)
    s = AdaptiveThresholdMoENeuronState()
    trace = zeros(n_steps)
    spikes = 0
    for t in 1:n_steps
        result = step!(s, I_ext; dt=dt)
        trace[t] = s.v
        if result isa Number && result > 0
            spikes += 1
        end
    end
    return trace, spikes
end

end # module AdaptiveThresholdMoeAccel
